drivers: flash: Add flash driver for MRAM

57300 opened the pull request:
Basic driver utilizing the flash API for NVM operations on the nRF54H20.

57300 committed:
drivers: flash: Add flash driver for MRAM

Basic driver utilizing the flash API for NVM operations on the nRF54H20.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

de-nordic commented on drivers/flash/soc_flash_nrf_mram.c:
```diff
@@ -0,0 +1,186 @@
+#define WRITE_BLOCK_SIZE MRAM_WORD_SIZE
+#endif
+
+#define PAGE_SIZE  CONFIG_SOC_FLASH_NRF_MRAM_LAYOUT_PAGE_SIZE
```
Can we do that via DTS?

nordicjm replied:
The mram dts instance does definitely have an erase and write block size set right? Because if not a lot of things will outright just not work, e.g. mcuboot

de-nordic replied:
We have to use pages for now as there is a lot of code that depends on int. Even when we provide ability to work with devices that do not require erase, which is in progress, there are still parts of system that do things by pages, like FCB for example. MCUboot is other example of that, as Jamie has mentioned.

57300 replied:
For the record, I retested this driver with FCB (as Settings backend in `samples/subsys/settings`) and it works OK, except IIRC it only works with write block size <= 4.

57300 replied:
Added `erase-block-size` property.

de-nordic commented on drivers/flash/soc_flash_nrf_mram.c:
```diff
@@ -0,0 +1,186 @@
+	barrier_dmem_fence_full();
+
+	if (WRITE_BLOCK_SIZE == MRAM_WORD_SIZE || (addr_end & MRAM_WORD_MASK) == 0) {
+		/* Our last operation was MRAM word-aligned, so we're done. */
```
Is this comment ok? Isn't that we return here if we have not yet written the last byte of the MRAM word, that we also leave here?

57300 replied:
To clarify, we enter here only if we ended exactly on the last byte of an MRAM word, hence:
`(addr_end & MRAM_WORD_MASK) == 0`.

If `WRITE_BLOCK_SIZE == MRAM_WORD_SIZE`, then the latter condition is already enforced by `validate_and_map_addr()`, so this is a compile-time optimization.

de-nordic commented on drivers/flash/soc_flash_nrf_mram.c:
```diff
@@ -0,0 +1,186 @@
+
+	LOG_DBG("read: %p:%zu", (void *)addr, len);
+
+	memcpy(data, (void *)addr, len);
```
What is read here if `nrf_mram_write` has not yet been commited? For example we have written WRITE_BLOCK_SIZE - 1 out of WRITE_BLOCK_SIZE bytes?

57300 replied:
An uncommitted write is buffered, so you'll usually read a stale value in this case.

de-nordic replied:
Fortunately these are not block devices... still a bummer.

de-nordic commented on drivers/flash/soc_flash_nrf_mram.c:
```diff
@@ -0,0 +1,186 @@
+ */
+static uintptr_t validate_and_map_addr(off_t offset, size_t len, bool must_align)
+{
+	if (unlikely(offset < 0 || offset + len > MRAM_SIZE)) {
```
I think that correct would be to do `offset < 0 || offset >= MRAM_SIZE || ((MRAM_SIZE - offset) < len)`, as this check `offset` against two constant boundaries we know, and then `len` against already verified range.
This also guards against `offset + len` overflowing to something that will be `< MRAM_SIZE`.

57300 replied:
Updated, thanks

de-nordic commented on drivers/flash/soc_flash_nrf_mram.c:
```diff
@@ -0,0 +1,186 @@
+		/* Address and length must both be multiples of MRAM_WORD_SIZE.
+		 * Since it's a power of two, we can check this with a bitmask.
+		 */
+		if (unlikely((addr | len) & MRAM_WORD_MASK)) {
```
This actually optimizes nicely, with no hidden compiler optimizations, to single branch. If I remember correctly this is three to five instructions on Arm.

57300 commented on drivers/flash/soc_flash_nrf_mram.c:
```diff
@@ -0,0 +1,186 @@
+#if defined(CONFIG_SOC_FLASH_NRF_MRAM_ONE_BYTE_WRITE_ACCESS)
+#define WRITE_BLOCK_SIZE 1
+#else
+#define WRITE_BLOCK_SIZE MRAM_WORD_SIZE
```
In other words, do you want me to `#define MIN_WRITE_BLOCK_SIZE 1` and check that the DTS value is divisible by it?

de-nordic replied:
No. What I mean is that if you do not set `CONFIG_SOC_FLASH_NRF_MRAM_ONE_BYTE_WRITE_ACCESS`, then you already have assumed that we are only allowed to write by some device specific write block size, as a minimal unit.
That is what device accepts, yet we allow to modify this by selecting something in DTS, so if device has write-block-size requirement then any value set by DTS should be multiply of device requirement.

In case if you set CONFIG_SOC_FLASH_NRF_MRAM_ONE_BYTE_WRITE_ACCESS, then you probably do not care what user sets there, any value is multiply of CONFIG_SOC_FLASH_NRF_MRAM_ONE_BYTE_WRITE_ACCESS.

The DTS can independently be used to modify behaviour of code, but should not violate what device can do; for example nrf52840 has 4 byte write-block-size, but user may want to have there 16, this is ok, but if user sets DTS to 2, that is something the device can not do.

57300 replied:
Having looked at other recent flash driver PRs, I decided to only use the MRAM word size hardcoded in the driver.

I don't think this is ideal, I think constants like this should be fetched from DTS, but there doesn't seem to be an established pattern for that in flash at the moment.

de-nordic replied:
Not really. New drivers actually start to use the DTS.
And yest soc_nrf_flash does that, which is pain in the lower parts if you try to adjust the write-block-size above what is hardcoded.
I will PR this change further or later, will not insist on this change here. Marking as resolved.

de-nordic reviewed (dismissed):
Looks OK, although I think that we are missing "erase-block-size" optional binding in DTS for the MRAM, and write-block-size being read from the DTS.
The second issue seems to be also on soc-nrf-flash driver, so will not insist on implementing.
The first one seems controversial, but if author will not pursue we can just do another PR with the change to binding and driver.

Laczen commented on drivers/flash/Kconfig.nrf_mram:
```diff
@@ -0,0 +1,40 @@
+	  This is supported in hardware, but it requires slightly more code to
+	  perform the operations safely.
+
+config SOC_FLASH_NRF_MRAM_LAYOUT_PAGE_SIZE
```
This should be moved to the dts or fixed in the driver. This will avoid any confusion:
* there are drivers where this isn't specified in the dts (because the blocks vary in size) in that case it is fixed by the driver.
* there are drivers where this is specified and then it is used for the page layout.

Adding it as a kconfig might result in applications on the same device that get stuck because they where compiled with different kconfig options.

In case where `ONE_BYTE_WRITE_ACCESS` is not selected it should also be a multiple of the `write-block-size`.

Laczen replied:
To be fair, also the `ONE_BYTE_WRITE_ACCESS` should also come from the dts (by setting `write-block-size` equal to 1. But I didn't check how this is done on other drivers.

57300 replied:
Alright, I'll move both Kconfigs to DT. I realize now that both `write-block-size` and `erase-block-size` seem to be treated as configurable anyway, which is strange, but I can play along. I'd like to come back to this at a future date.

57300 committed:
dts: nordic: nrf54h20_enga: Add erase-block-size to mram1x

The MRAM has no concept of erase blocks or pages, so this is treated as
driver configuration.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

Laczen reviewed (approved):
LGTM
